src/main/java/com/auth0/utils/tokens/SignatureVerifier.java

```diff
@@ -0,0 +1,83 @@
+package com.auth0.utils.tokens;
+
+import com.auth0.exception.IdTokenValidationException;
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.exceptions.AlgorithmMismatchException;
+import com.auth0.jwt.exceptions.JWTDecodeException;
+import com.auth0.jwt.exceptions.JWTVerificationException;
+import com.auth0.jwt.exceptions.SignatureVerificationException;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.auth0.jwt.interfaces.JWTVerifier;
+import com.auth0.utils.Asserts;
+
+/**
+ * Represents the verification for an ID Token's signature used when validating an ID token.
+ */
+public abstract class SignatureVerifier {
+
+    private final JWTVerifier verifier;
+
+    /**
+     * Get a {@code SignatureVerifier} for use when validating an ID token signed using the HS256 signing algorithm.
+     *
+     * @param secret the client's secret to use when validating the token's signature.
+     * @return a {@code SignatureVerifier} for use with tokens signed using the HS256 signing algorithm.
+     */
+    public static SignatureVerifier forHS256(String secret) {
+        return new HS256SignatureVerifier(secret);
+    }
+
+    /**
+     * Get a {@code SignatureVerifier} for use when validating an ID token signed using the RS256 signing algorithm.
+     * Callers should provide an implementation of the {@linkplain PublicKeyProvider} to provide the public key used
+     * to verify the ID token's signature.
+     *
+     * @see PublicKeyProvider
+     * @param publicKeyProvider an implementation of {@linkplain PublicKeyProvider} to get the public key.
+     * @return a {@code SignatureVerifier} for use with tokens signed using the RS256 signing algorithm.
+     */
+    public static SignatureVerifier forRS256(PublicKeyProvider publicKeyProvider) {
+        return new RS256SignatureVerifier(publicKeyProvider);
+    }
+
+    /**
+     * Creates a new JWT Signature Verifier. Used by internal implementations to create concrete verifiers.
+     *
+     * @param verifier the instance that knows how to verify the signature. Must not be {@code null}.
+     */
+    SignatureVerifier(JWTVerifier verifier) {
+        Asserts.assertNotNull(verifier, "verifier");
```

Comment:
I think this will never be the case? Since this class is the only one creating them from static methods. In any case, the null checks could be when receiving the client secret and key provider instances.

Reply:
You are correct, since we are the only one creating them and the constructor isn't public. The motivation was to ensure that future (internal) implementations would need to either conform to this, or explicitly remove the non-null requirement.

Reply:
Let's keep it then, it doesn't hurt 👍